### PR TITLE
Added API that returns information about Cadence deployment

### PR DESCRIPTION
**Pull request by anish531213**

### CODE REVIEW

File: service/frontend/workflowHandler.go [hunk 1]
```diff
@@ -3512,6 +3512,17 @@ func (wh *WorkflowHandler) ReapplyEvents(
 	return nil
 }
 
+// GetCadenceDeploymentInfo return information about cadence deployment
+func (wh *WorkflowHandler) GetCadenceDeploymentInfo(ctx context.Context) (resp *gen.DeploymentInfo, err error) {
```
Comment (anish531213) [inline service/frontend/workflowHandler.go]: Right now I am not checking for any error. Any suggestion on errors that needs to be handled?

File: idl/github.com/uber/cadence/cadence.thrift [hunk 2]
```diff
@@ -620,4 +620,9 @@ service WorkflowService {
       5: shared.ServiceBusyError serviceBusyError,
       6: shared.EntityNotExistsError entityNotExistError,
     )
+
+  /**
+  * GetCadenceDeploymentInfo returns information about cadence deployment
+  **/
+  shared.DeploymentInfo GetCadenceDeploymentInfo()
```
Comment (mfateev) [inline idl/github.com/uber/cadence/cadence.thrift]: What is Deployment? 
May be CadenceServiceInfo? Or CadenceClusterInfo?

File: idl/github.com/uber/cadence/shared.thrift [hunk 3]
```diff
@@ -1591,4 +1591,16 @@ struct ReapplyEventsRequest{
...
+// SupportedClientVersions contains the support versions for client library
+struct SupportedClientVersions{
+  10: optional string gosdk
+  20: optional string javasdk
+  30: optional string cli
+}
+
+// DeploymentInfo contains information about cadence deployment
+struct DeploymentInfo{
+  10: optional SupportedClientVersions supportedClientVersions
```
Comment (mfateev) [inline idl/github.com/uber/cadence/shared.thrift]: No cluster membership information?

File: service/frontend/workflowHandler.go (hunk 1 above)
Comment (mfateev) [reply service/frontend/workflowHandler.go]: Ideally this call should return some flag that indicated that cluster is healthy and ready to take traffic. It is needed for example during docker startup to wait for service to start by polling this function.

File: idl/github.com/uber/cadence/shared.thrift (hunk 3 above)
Comment (anish531213) [reply idl/github.com/uber/cadence/shared.thrift]: I am starting with just one at the beginning and add to it on next PR(s).

File: service/frontend/workflowHandler.go (hunk 1 above)
Comment (wxing1292) [reply service/frontend/workflowHandler.go]: health check API and cluster info API should be 2 different APIs

Comment (wxing1292) [reply service/frontend/workflowHandler.go]: i would also suggest making this API `GetClusterInfo`

File: idl/github.com/uber/cadence/cadence.thrift [hunk 4]
```diff
@@ -620,4 +620,9 @@ service WorkflowService {
       5: shared.ServiceBusyError serviceBusyError,
       6: shared.EntityNotExistsError entityNotExistError,
     )
+
+  /**
+  * GetCadenceClusterInfo returns information about cadence cluster
+  **/
+  shared.CadenceClusterInfo GetCadenceClusterInfo()
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply idl/github.com/uber/cadence/cadence.thrift]: GetClusterInfo

[the remark this answers is not in the document]
Comment (wxing1292) [reply idl/github.com/uber/cadence/cadence.thrift]: also shared.ClusterInfo

File: idl/github.com/uber/cadence/shared.thrift [hunk 5]
```diff
@@ -1591,4 +1591,16 @@ struct ReapplyEventsRequest{
   10: optional string domainName
   20: optional WorkflowExecution workflowExecution
   30: optional DataBlob events
+}
+
+// SupportedClientVersions contains the support versions for client library
+struct SupportedClientVersions{
+  10: optional string gosdk
+  20: optional string javasdk
+  30: optional string cli
```
Comment (wxing1292) [inline idl/github.com/uber/cadence/shared.thrift]: why there is a cli version?
cli it self use go

File: service/frontend/workflowHandler.go [hunk 6]
```diff
@@ -3512,6 +3512,21 @@ func (wh *WorkflowHandler) ReapplyEvents(
...
+func (wh *WorkflowHandler) GetCadenceClusterInfo(
+	ctx context.Context,
+) (resp *gen.CadenceClusterInfo, err error) {
+	defer log.CapturePanic(wh.GetLogger(), &err)
+
+	return &gen.CadenceClusterInfo{
+		SupportedClientVersions: &gen.SupportedClientVersions{
+			Gosdk:   common.StringPtr(fmt.Sprintf("<=%v", client.SupportedGoSDKVersion)),
+			Javasdk: common.StringPtr(fmt.Sprintf("<=%v", client.SupportedJavaSDKVersion)),
+			Cli:     common.StringPtr(fmt.Sprintf("<=%v", client.SupportedJavaSDKVersion)),
```
Comment (wxing1292) [inline service/frontend/workflowHandler.go]: why `<=`

File: service/frontend/workflowHandler.go [hunk 7]
```diff
@@ -3512,6 +3512,21 @@ func (wh *WorkflowHandler) ReapplyEvents(
 	return nil
 }
 
+// GetCadenceClusterInfo return information about cadence deployment
+func (wh *WorkflowHandler) GetCadenceClusterInfo(
+	ctx context.Context,
+) (resp *gen.CadenceClusterInfo, err error) {
+	defer log.CapturePanic(wh.GetLogger(), &err)
+
```
Comment (wxing1292) [inline service/frontend/workflowHandler.go]: also include the throttling

Comment (anish531213) [reply service/frontend/workflowHandler.go]: I didn't understand what you meant by including the throttling. Can you point me to an example?

File: idl/github.com/uber/cadence/shared.thrift [hunk 8]
```diff
@@ -1591,4 +1591,15 @@ struct ReapplyEventsRequest{
   10: optional string domainName
   20: optional WorkflowExecution workflowExecution
   30: optional DataBlob events
+}
+
+// SupportedClientVersions contains the support versions for client library
+struct SupportedClientVersions{
+  10: optional string gosdk
+  20: optional string javasdk
```
Comment (wxing1292) [inline idl/github.com/uber/cadence/shared.thrift]: `goSdk` and `javaSdk` i guess

File: service/frontend/workflowHandler.go [hunk 9]
```diff
@@ -3512,6 +3512,21 @@ func (wh *WorkflowHandler) ReapplyEvents(
 	return nil
 }
 
+// GetClusterInfo return information about cadence deployment
+func (wh *WorkflowHandler) GetClusterInfo(
+	ctx context.Context,
+) (resp *gen.ClusterInfo, err error) {
+	defer log.CapturePanic(wh.GetLogger(), &err)
+	// Count the request in the RPS, but we still accept it even if RPS is exceeded
+	wh.allow(nil)
```
Comment (yycptt) [inline service/frontend/workflowHandler.go]: Please check the error returned by this function call.

Comment (anish531213) [reply service/frontend/workflowHandler.go]: fixed

**Comment by meiliang86** [comment]
@anish531213 We will probably need a corresponding command in CLI as well.

**Pull request merged**